source/extensions/priv/server/ntds_jet.c

```diff
@@ -0,0 +1,312 @@
+#include "precomp.h"
+
+JET_ERR engine_shutdown(jetState *ntdsState){
+	JET_ERR shutdownStatus;
+	shutdownStatus = JetCloseDatabase(ntdsState->jetSession, ntdsState->jetDatabase, (JET_GRBIT)NULL);
+	if (shutdownStatus != JET_errSuccess){
+		return shutdownStatus;
+	}
+	shutdownStatus = JetDetachDatabase(ntdsState->jetSession, ntdsState->ntdsPath);
+	if (shutdownStatus != JET_errSuccess){
+		return shutdownStatus;
+	}
+	shutdownStatus = JetEndSession(ntdsState->jetSession, (JET_GRBIT)NULL);
+	if (shutdownStatus != JET_errSuccess){
+		return shutdownStatus;
+	}
+	shutdownStatus = JetTerm(ntdsState->jetEngine);
+	free(ntdsState);

```

Comment:
ah! This might be the place where things are freed. Perhaps it'd be better to put both the alloc/free in the same locations? That at least then implies who is responsible for managing the resource. Maybe `engine_startup` should allocate `ntdsState`?